test/ui-test/testSuites/suite_statusLogin/shared/steps/steps.py

```diff
@@ -39,23 +45,43 @@ def step(context):
     # Check valid process behavior:
     loginProcess.execute_process(True)
 
+
 @When("the user tries to login with invalid credentials")
 def step(context):
     loginProcess = context.userData['process']
     
     # Check invalid process behavior:
     loginProcess.execute_process(False)
 
+
 @Then("the user is able to login to Status Desktop application")
 def step(context):
     get_process_result(context)
+
     
 @Then("the user is NOT able to login to Status Desktop application")
 def step(context):
     get_process_result(context)
-  
+    
+
+@Given("A first time user lands on the status desktop and generates new key") 
```

Comment:
Is it really a `Given` or a `When`? These are actions that are part of the process of login. Maybe the `Given` here is setting up the user profile we want to test (creating the user account to interact with it, like storing the user name, password, maybe the language we want to test)..

I think that `Given` statements should be mainly to create needed objects in the test and store them here:
`context.userData['object1']`
And execute the needed actions to point the application to the stage you need and really want to test.

Reply:
Given, When, Then, And, do not matter what structure you use. 
Let's not add rules to this. 
It makes it very restrictive and harder to work with.

I avoid storing data in contexts. I prefer to have all data in the gherkin file.

Given I have 'A"
When I use 'A'
Then 'A' is not 'B'

I only store data (In your context) as a last resort.
And when that happens I will know why I had to.

Reply:
It is better to store data in the context instead of having a **global property** anywhere. That's exactly the propose of `context.userData`! It is not related to having written the property in a the gherkin file or in another place it is about how you manage the data itself. `context.userData` is a specific object that stores all context data.